### PR TITLE
Optimize EREGOL check

### PR DESCRIPTION
**Pull request by mandolaerik**
This speeds up the check by around 50x for large arrays.

### CODE REVIEW

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631023

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631022

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631573

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631581

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631572

**Comment by syssimics** [comment]
TeamCity build completed (FAILURE): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631580

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631678

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631677

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631958

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9631957

**Review by lwaern-intel — CHANGES_REQUESTED** [review]
Good stuff. The implementation looks reasonable enough, but I'd like some more thorough tests.

File: py/dml/structure.py [hunk 1]
```diff
@@ -1987,15 +1989,68 @@ def get_group_registers(group, indexvars):
 
     return regs
 
+def param_linear_int(param):
+    '''If possible, return tuple of coefficients (i0, i1, .., iN, k) such
+    that expr can be equivalently expressed as (i0*v0 + i1*v1 + ... +
```
Comment (JonatanWaern) [inline py/dml/structure.py:1994]: This comment confused me a fair bit, mostly because what 'expr' is is not obvious at point of reading. Consider "expr of param" or similar

File: test/1.4/lib/T_largearray.dml [hunk 2]
```diff
@@ -12,8 +12,8 @@ bank b {
     register a[i < 10000] @ 50000 + (4 * i) is read_only;
     register b[i < 10000] @ 90000 + (4 * i) is read_only;
 
-    register c[i < 10000] @ 150000 + (8 * i) is read_only;
-    register d[i < 10000] @ 150004 + (8 * i) is read_only;
+    register c[i < 100000] @ 1500000 + (8 * i) is read_only;
+    register d[i < 100000] @ 1500004 + (8 * i) is read_only;
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply test/1.4/lib/T_largearray.dml:16]: The penalty for the 100000 size is around 0.1 s. The biggest cost is now is the 3M register dispatches triggered by read(), takes around 1s. Perhaps we should reduce that loop by an order of magnitude.

I want the test's running time to be close to annoying, so that a significant regression will be noticeable when running the test suite manually.

File: py/dml/structure.py [hunk 3]
```diff
@@ -1987,15 +1989,68 @@ def get_group_registers(group, indexvars):
...
+        elif isinstance(expr, ctree.IntegerConstant):
+            return (0,) * param.dimensions + (expr.value,)
+        elif isinstance(expr, IndexVar):
+            return expr.variables
+        elif (isinstance(expr, ctree.Cast) and expr.type.is_int
+              and expr.type.bits == 64):
+            return expr_linear_int(expr.expr)
+        raise NotLinear()
+
+    expr = param.get_expr(indices)
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/structure.py:2039]: Good point, I should probably duck your question by converting any raised `DMLError` to `return None`. The function's task is to search for a certain pattern, and an erratic expression doesn't have that pattern.

File: py/dml/structure.py (hunk 1 above)
Comment (mandolaerik) [reply py/dml/structure.py:1994]: Good catch! I initially wrote the comment on the inner function, and forgot to update it when moving it to the outer function.

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9637071

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9637070

File: py/dml/structure.py [hunk 4]
```diff
@@ -1987,15 +1989,69 @@ def get_group_registers(group, indexvars):
...
+    try:
+        return tuple(c & ((1 << 64) - 1) for c in expr_linear_int(expr))
+    except (NotLinear, DMLError):
```
Comment (lwaern-intel) [inline py/dml/structure.py:2043]: Isn't this catch on `DMLError` in the wrong place? Shouldn't it be on the `param.get_expr` line?
(Un)fortunately, Python isn't a lazy language :P

Comment (mandolaerik) [reply py/dml/structure.py]: Right! Which also reveals that I didn't write a test for this corner. Done now.

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9637526

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9637525

**Review by lwaern-intel — COMMENTED** [review]
Looks good. Ready to approve once my final question is answered.

File: py/dml/structure.py [hunk 5]
```diff
@@ -1987,15 +1989,72 @@ def get_group_registers(group, indexvars):
...
+
 def sort_registers(bank):
     # TODO: partition_registers is expensive, and all it gives in 1.4
     # is to check for errors such as EREGOL. It would be better with a
     # function that only checks for those problems, but which yields a
     # warning instead. That way, we can naturally allow the
     # (expensive) check to be skipped by disabling the warning. See US1610.
-    (bank.mapped_registers, bank.unmapped_registers,
-     bank.numbered_registers) = explode_registers(bank)
     if dml.globals.dml_version == (1, 2):
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/structure.py:2055]: Yes, integer semantics is predictable in 1.4. In 1.2, it's hard to know whether `$i + 3` implicitly also does `% 0x100000000`.

[the remark this answers is not in the document]
Comment (lwaern-intel) [reply py/dml/structure.py:2055]: I see!

**Review by lwaern-intel — APPROVED** [review]
Friendly reminder to squash, rebase, and merge.

**Review by lwaern-intel — APPROVED** [review]

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9703510

**Comment by syssimics** [comment]
TeamCity build completed (SUCCESS): https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=9703509
